### PR TITLE
Add py-atldld

### PR DESCRIPTION
**Pull request by Stannislav**
- Repo: https://github.com/BlueBrain/Atlas-Download-Tools
- PyPI: https://pypi.org/project/atldld

Other packages that had to be added
- `py-opencv-python`

My first time adding something to spack, so any feedback is very welcome!

### CODE REVIEW

**Comment by Stannislav** [comment]
I'm not sure I did everything right in `deploy/environments/applications.yaml`, I tried to mimick what I saw elsewhere...

On BB5 this works
```
$ spack install py-atldld
$ spack load py-atldld
```

But
```
$ spack module tcl refresh -y
==> Warning: Skipping whitelisted module ...
...
$ module load atldld
ModuleCmd_Load.c(213):ERROR:105: Unable to locate a modulefile for 'atldld'
```
Not sure if this is expected, probably I'm missing something. Please do point me to the flaws.

File: deploy/environments/applications.yaml [hunk 1]
```diff
@@ -138,6 +140,7 @@ spack:
     - py-morph-tool@2.8.0%gcc ^neuron%intel
     - py-neurodamus+all_deps
     - py-notebook
+    - py-opencv-python
```
Comment (matz-e) [inline deploy/environments/applications.yaml]: Do we want to build this explicitly? It should already be in the dependency tree for `py-atldld`

Comment (Stannislav) [reply deploy/environments/applications.yaml]: It's probably enough to have it as a dependency of `py-atldld` for now. Is it just a matter of removing this line, or is there something else I have to do?

Comment (matz-e) [reply deploy/environments/applications.yaml]: Just removing the line should be good enough.

**Comment by matz-e** [comment]
> Not sure if this is expected, probably I'm missing something. Please do point me to the flaws.

Think you forgot a `py-`?

```
matwolf in bbpv2 in ~ via  v14.13.0 via 🐍 v2.7.5
❯ unset MODULEPATH

matwolf in bbpv2 in ~ via  v14.13.0 via 🐍 v2.7.5
❯ . /gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/pulls/1286/config/modules.sh

matwolf in bbpv2 in ~ via  v14.13.0 via 🐍 v2.7.5
❯ module load unstable

matwolf in bbpv2 in ~ via  v14.13.0 via 🐍 v2.7.5
❯ module av py-atldld

------------------- /gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/pulls/1286/modules/applications/2021-09/tcl/linux-rhel7-x86_64 --------------------
py-atldld/0.2.2

matwolf in bbpv2 in ~ via  v14.13.0 via 🐍 v2.7.5
❯ module load py-atldld
Autoloading python/3.8.3
Autoloading ffmpeg/4.2.2

matwolf in bbpv2 in ~ via  v14.13.0 via 🐍 v3.8.3
❯
```
seems to work for me. Now I'm clueless about the software, so I'm a bad tester :)

**Comment by Stannislav** [comment]
> seems to work for me. 

Must've been `$MODULEPATH` and sourcing of `modules.sh`, I wasn't aware of this. And yes, also `py-`:
```
$ module load unstable
$ module av py-atldld
<nothing>

$ echo $MODULEPATH
/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/applications/2021-01-06/modules/tcl/linux-rhel7-x86_64:/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/libraries/2021-01-06/modules/tcl/linux-rhel7-x86_64:/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/modules/all:/gpfs/bbp.cscs.ch/apps/compilers/modules/tcl/linux-rhel7-x86_64:/gpfs/bbp.cscs.ch/apps/tools/modules/tcl/linux-rhel7-x86_64:/gpfs/bbp.cscs.ch/apps/modules/teams:/gpfs/bbp.cscs.ch/ssd/apps/nix/modules:/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/modules/compilers/2021-09/tcl/linux-rhel7-x86_64:/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/modules/externals/2021-09/tcl/linux-rhel7-x86_64:/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/modules/libraries/2021-09/tcl/linux-rhel7-x86_64:/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/modules/applications/2021-09/tcl/linux-rhel7-x86_64
$ unset MODULEPATH
$ . /gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/pulls/1286/config/modules.sh
$ module load unstable
$ module av py-atldld

------------------------------------------------------------------ /gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/pulls/1286/modules/applications/2021-09/tcl/linux-rhel7-x86_64 -------------------------------------------------------------------
py-atldld/0.2.2
$ module load py-atldld
Autoloading python/3.8.3
Autoloading ffmpeg/4.2.2
```

**Comment by Stannislav** [comment]
> Now I'm clueless about the software, so I'm a bad tester :)

I ran the unit tests on a local clone (after using the modules loaded like in my previous comment), seems to work:
```
$ cd ../atldld
$ spack load /l3g3buc /cmoxnep  # py-pytest and py-pytest-cov
$ spack load py-pytest-mock
$ pytest tests
=========================================================================================================== test session starts ============================================================================================================
platform linux -- Python 3.8.3, pytest-5.3.4, py-1.8.0, pluggy-0.13.0
rootdir: /gpfs/bbp.cscs.ch/home/sschmidt/code/atldld, inifile: tox.ini, testpaths: tests
plugins: mock-1.11.1, cov-2.8.1
collected 378 items                                                                                                                                                                                                                        

tests/test_base.py ................................................................................................................................................................................................................. [ 55%]
......................................................................................................                                                                                                                               [ 82%]
tests/test_sync.py .........................                                                                                                                                                                                         [ 88%]
tests/test_utils.py .........................................                                                                                                                                                                        [ 99%]
tests/test_version.py .                                                                                                                                                                                                              [100%]

===================================================================================================== 378 passed in 372.24s (0:06:12) ======================================================================================================
```

File: deploy/environments/applications.yaml (hunk 1 above)
Comment (matz-e) [reply deploy/environments/applications.yaml]:
```suggestion
```
This and we're good to merge.

Comment (Stannislav) [reply deploy/environments/applications.yaml]: OK, fixed in 39c7d09af5

Comment (Stannislav) [reply deploy/environments/applications.yaml]: And having two entries for `py-atldld` in `applications.yaml` is correct? I'm not so sure what each of them is for.

**Review by matz-e — APPROVED** [review]

**Comment by Stannislav** [comment]
So what's the merging policy, should I squash-merge myself?

@lguyot Hi Luc, I don't know if you want to have a look. This will be either directly part of the atlas pipelines, or part of dependencies for Atlas Annotation Alignment / Atlas Interpolation, which I'll try to deploy next. Do I also need to push something to `hpc/jenkins-pipeline`?

**Comment by matz-e** [comment]
> So what's the merging policy, should I squash-merge myself?

yes, you have the honor :)

**Comment by lguyot** [comment]
@Stannislav Thanks for notifying me. I presume the new project of this new package deprecates the former `atlalign.allen.download_dataset` which is currently used by https://bbpgitlab.epfl.ch/nse/gene-expression-volume.
Depending on user needs, I'll update `gene-expression-volume` after discussion with @mgeplf.

> This will be either directly part of the atlas pipelines, or part of dependencies for Atlas Annotation Alignment / Atlas Interpolation.

Are you dismembering and re-constructing `atlalign` in several new projects? Your next steps seem also to relate to `gene-expression-volume` as the latter interpolate 2D slice images to create a 3D volume, after slice realignment via non-ML (intensity based registration) and ML methods.

**Comment by Stannislav** [comment]
@lguyot `atlalign` is staying, however, the functionality related to the data download and interaction with the Allen servers (the `allen` module) has been migrated to the new `atldld` module that's deployed here. In short, `atldld` is the common ground for data download/preparation for the following 3 projects:
* `atlalign` "Atlas Alignment": alignment of the ISH data
* `atlannot` "Atlas Annotation": alignment of the atlas annotations
* `atlinter` "Atlas Interpolation": interpolation of the ISH data

Maybe also ping @FrancescoCasalegno on Slack to stay in the loop for this (maybe you already are), I think there's going to be some follow up on this in October.

**Pull request merged**